11-payment-encoding.md

```diff
@@ -509,6 +533,44 @@ Breakdown:
   * `6c6e626332306d0b25fe64570d0e496dbd9f8b0d000dbb44824f751380da37c6dba89b14f6f92047d63f576e304021a00008101820283038404800081018202830384048000810182028303840480810243500c318a1e0a628b34025e8c9019ab6d09b64c2b3c66a693d0dc63194b02481931000` hex of data for signing (prefix + data after separator up to the start of the signature)
   * `399a8b167029fda8564fd2e99912236b0b8017e7d17e416ae17307812c92cf42` hex of SHA256 of the preimage
 
+> ### Please send $30 for coffee beans to the same peer, which supports features 1 and 9
+> lnbc25m1pvjluezpp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqdq5vdhkven9v5sxyetpdees9qzsze992adudgku8p05pstl6zh7av6rx2f297pv89gu5q93a0hf3g7lynl3xq56t23dpvah6u7y9qey9lccrdml3gaqwc6nxsl5ktzm464sq73t7cl
```

Comment:
the current test vector (at least in my impl) produces an incomplete group when converting the feature bit data from base32 -> base256, i suspect we're disagreeing on a base 256 <-> base32 encoding somewhere?

for this i get:
```
lnbc25m1pvjluezpp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqdq5vdhkven9v5sxyetpdees9qyqgpq6w0yzlsl2aeqgg2v60eldr2cl6z59rcmlj255xz5pppewqw0mtpx5tywjnk7jna83jldf85z23k4zx2whl9q7jlzqge48ur3fk5ewcgqn5nkkk
```

with the vectors i arrived at i am able to encode and decode in both directions, but not with the ones in the proposal. @t-bast were you able to reproduce the original vectors?

Reply:
Yes, with my implementation I end up with the same result as rusty's.
Indeed your vector disagrees on the `9` field, so it looks like this is linked to the bits encoding...could it be because you're not properly truncating the number of bits to a multiple of 5?

Reply:
currently applying the same transformation as all other fields, can you confirm that the raw bytes before converting from base32 to base256 are `[0b10000, 0b10]`?

Reply:
Okay I did some digging, here are my findings.

The encoding of feature bits 1 and 9 uses 16 bits in base 256:
```
[0b00000010, 0b00000010]
```

Converting this to base32 requires `ceil(16/5) = 4` characters, and translates to:
```
[0b00000, 0b01000, 0b00001, 0b0====]
```
where `=` are padding bits.

I was able to reproduce this using sipa's [python `convertbits` method](https://paste.ubuntu.com/p/xFK9SsPYvt/).

The input `[0b10000, 0b00010]` fails to convert due to the same invalid group error I encountered. OTOH, `[0b00000, 0b01000, 0b00001, 0b00000]` (the encoding used in the vector I provided) successfully converts to `[0b00000010, 0b00000010]`.

Here's a paste of the experiment https://paste.ubuntu.com/p/xFK9SsPYvt/|

I'm not seeing how a valid vector could only have two base32 characters, i.e. 10 out of the 16 base256 bits.

IINM, this should be:
```
 * `9`: features
    * `qy` `data_length` (`q` = 0, `y` = 4; 0 * 32 + 4 == 4)
    * `qgpq`: 0b00000|01000|00001|00000
```

Reply:
FWIW I also reimplemented our convertbits method identically to the python one to rule out any possible bugs in our implementation and I get the same results. I looked about both eclair's and CL's converbit methods and they appear correct to me, so if the vector _is_ correct that means we are disagreeing somewhere else.

Reply:
Okay I think I understand what’s happening. I was incorrectly assuming that we would be reusing the wire encoding and converting to base32, rather than encoding to base32 directly. 

IMO this additional complexity and code duplication isn’t really warranted—it saves at most 2 characters and is dwarfed by things like route hints..

It’s also quite a departure from virtually every other complex field that already uses convertbits to get in and out of base32

Reply:
> I was incorrectly assuming that we would be reusing the wire encoding and converting to base32, rather than encoding to base32 directly.

Got it, I understand why your results diverged then. So you're able to reproduce the test vector now?
I think it makes sense to go directly to base32, it's clearer and easier to debug (once you know that's what's done :p).

> IMO this additional complexity and code duplication isn’t really warranted—it saves at most 2 characters and is dwarfed by things like route hints..

It's not code duplication for us, I think this is highly dependent on how that was implemented. Payment request encoding has been separate from message encoding from the start, hasn't it? I'll let rusty provide more details on that as I wasn't there when this was initially defined.